flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeSubpartitionReader.java

```diff
@@ -19,154 +19,180 @@
+            if (buffer == null) {
+                return null;
+            }
+
+            if (buffer.isBuffer()) {
+                --dataBufferBacklog;
+            }
 
-        Buffer buffer = buffersRead.poll();
-        if (buffer == null) {
-            return null;
+            Buffer lookAhead = buffersRead.peek();
+            return BufferAndBacklog.fromBufferAndLookahead(
+                    buffer,
+                    lookAhead == null ? Buffer.DataType.NONE : lookAhead.getDataType(),
+                    dataBufferBacklog,
+                    sequenceNumber++);
         }
+    }
 
-        if (buffer.isBuffer()) {
-            --dataBufferBacklog;
+    private void addBuffer(Buffer buffer) {
+        boolean notifyAvailable;
+        synchronized (lock) {
+            if (isReleased) {
+                buffer.recycleBuffer();
+                return;
+            }
+
+            notifyAvailable = buffersRead.isEmpty();
+
+            buffersRead.add(buffer);
+            if (buffer.isBuffer()) {
+                ++dataBufferBacklog;
+            }
         }
 
-        final Buffer lookAhead = buffersRead.peek();
+        if (notifyAvailable) {
+            notifyDataAvailable();
+        }
+    }
+
+    boolean readBuffers(Queue<MemorySegment> buffers, BufferRecycler recycler) throws IOException {
+        while (!buffers.isEmpty()) {
+            MemorySegment segment = buffers.poll();
 
-        return BufferAndBacklog.fromBufferAndLookahead(
-                buffer,
-                lookAhead == null ? Buffer.DataType.NONE : lookAhead.getDataType(),
-                dataBufferBacklog,
-                sequenceNumber++);
+            Buffer buffer;
+            try {
+                if ((buffer = fileReader.readCurrentRegion(segment, recycler)) == null) {
+                    buffers.add(segment);
+                    break;
+                }
+            } catch (Throwable throwable) {
+                buffers.add(segment);
+                throw throwable;
+            }
+            addBuffer(buffer);
+        }
+        return fileReader.hasRemaining();
     }
 
-    void readBuffers() throws IOException {
-        // we do not need to recycle the allocated segment here if any exception occurs
-        // for this subpartition reader will be released so no resource will be leaked
-        MemorySegment segment;
-        while ((segment = readBuffers.poll()) != null) {
-            Buffer buffer = fileReader.readBuffer(segment, this);
-            if (buffer == null) {
-                readBuffers.add(segment);
-                break;
+    void fail(Throwable throwable) {
+        checkArgument(throwable != null, "Must be not null.");
+
+        synchronized (lock) {
+            if (isReleased) {
+                return;
+            }
+
+            if (failureCause == null) {
+                failureCause = throwable;
             }
-            buffersRead.add(buffer);
         }
+
+        releaseAllResources();
```

Comment:
Could we extract a new `releaseResourceUnsafe(@Nullable Throwable throawable)` method and call it in both `fail()` and `releaseAllResource()` method ? Then we do not need to run duplicate `if (isReleased) {..}`